File: 0011-container-with-most-water/0011-container-with-most-water.java
class Solution {
    // static {
    //     for (int i = 0; i < 60; i++) {
    //         maxArea(new int[]{0, 0});
    //     }
    // }
    public static int maxArea(int[] height){
        int maxArea = 0, left = 0 , right = height.length - 1;
        while (left < right) {
            int h = Math.min(height[right],height[left]);
            maxArea = Math.max(maxArea, (right - left) * h);

            while(left < right && h >=height[left]) left++;
            
            while(left < right && h >= height[right]) right--;
        } 
        return maxArea;
    }
}

File: 0011-container-with-most-water/0011-container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        n = len(height)
        l = 0 
        r = n-1
        ans = 0 
        while l < r:
            h = min(height[l],height[r])
            ans = max(ans,h*(r-l))

            while l<r and h>=height[l] :
                l+=1
                
            while l<r and h>=height[r] :
                r-=1
        return ans